---
title: 'Shazam: Tuning clonal assignment thresholds with nearest neighbor distances
  neighbor'
author: "Namita Gupta"
date: '`r Sys.Date()`'
output:
  pdf_document:
    dev: pdf
    fig_height: 4
    fig_width: 7.5
    highlight: pygments
    toc: yes
  html_document:
    fig_height: 4
    fig_width: 7.5
    highlight: pygments
    theme: readable
    toc: yes
geometry: margin=1in
fontsize: 11pt
vignette: >
  %\VignetteEngine{knitr::rmarkdown} 
  %\VignetteIndexEntry{Distance to nearest neighbor} 
  %\usepackage[utf8]{inputenc}
---

Estimating the optimal distance threshold for partitioning clonally related 
sequences is accomplished by calculating the distance from each sequence in the 
data set to its nearest neighbor and finding the break point in the resulting 
bi-modal distribution that separates clonally related from unrelated sequences. 
This is done via the following steps:

1. Calculate the nearest neighbor distances for each sequence.
2. Generate a histogram of the nearest neighbor distances and inspect for the 
   threshold separating the two modes.

## Example data

A small example Change-O database is included in the `alakazam` package. 
Calculating the nearest neighbor distances requires the following 
fields (columns) to be present in the Change-O database: 

* `SEQUENCE_ID`
* `V_CALL`
* `J_CALL`
* `JUNCTION`
* `JUNCTION_LENGTH`

```{r, eval=TRUE, warning=FALSE, message=FALSE}
# Subset example data to one sample
library(shazam)
data(ExampleDb, package="alakazam")
db <- subset(ExampleDb, SAMPLE == "-1h")
```

## Calculate nearest neighbor distances

The function for calculating distance between every sequence and its nearest
neighbor takes a few parameters to adjust how the distance is measured. If a 
genotype has been inferred using the methods in the `tigger` package, and a 
`V_CALL_GENOTYPED` field has been added to the database, then this column may be 
used instead of the default `V_CALL` column by specifying the `vCallColumn` 
argument. This will allows the more accurate V call from `tigger` to be used for 
grouping of the sequences. Furthermore, for more leniency toward ambiguous 
V(D)J segment calls, the parameter `first` can be set to `FALSE`. Setting 
`first=FALSE` will use the union of all possible genes to group sequences, rather 
than the first gene in the field. The `model` parameter determines which 
underlying SHM model is used to calculate the distance. The default model is 
`hs1f`, a human Ig-specific single nucleotide model similar to a 
transition/transversion model (Yaari et al, 2013). Other options 
include nucleotide Hamming distance (`ham`), amino acid Hamming distance (`aa`), 
single nucleotide (`m1n`) and 3-mer (`m3n`) mouse models (Smith et al, 1996), 
and a 5-mer model inferred from human data (`hs5f`) (Yaari et al, 2013).
For models that are not symmetric (e.g., distance from A to B is not equal to the
distance from B to A), there is a `symmetry` parameter that allows the user to 
specify taking the average or the minimum of the two distances to determine the
overall distance.

```{r, eval=TRUE, warning=FALSE}
# Use hs1f model and normalize by junction length
dist_hs1f <- distToNearest(db, model="hs1f", first=FALSE, normalize="length", 
                           nproc=1)

# Use genotyped V assignments and 5-mer model
dist_hs5f <- distToNearest(db, vCallColumn="V_CALL_GENOTYPED", model="hs5f", 
                           first=FALSE, normalize="none", nproc=1)
```

## Generate histogram

The primary use of the distance to nearest calculation in the Change-O pipeline 
is to determine the optimal threshold for separating clonally related sequences 
(represented by sequences with "near"" neighbors) from singletons (sequences 
without "near"" neighbor), which show up as two modes in a histogram.

```{r, eval=TRUE, warning=FALSE, fig.width=7}
# Generate m1n histogram
library(ggplot2)
p1 <- ggplot() + theme_bw() + 
    ggtitle("Distance to nearest: hs1f") + xlab("distance") +
    geom_histogram(data=dist_hs1f, aes(x=DIST_NEAREST), binwidth=0.025, 
                   fill="steelblue", color="white") +
    geom_rug(aes(x=seq(0.0, 0.5, 0.1)), color="firebrick")
plot(p1)
```

In this example, the length normalized `hs1f` model distance threshold would be 
set to a value near 0.1.

```{r, eval=TRUE, warning=FALSE, fig.width=7}
# Generate hs5f histogram
p2 <- ggplot() + theme_bw() + 
    ggtitle("Distance to nearest: hs5f") + xlab("distance") +
    geom_histogram(data=dist_hs5f, aes(x=DIST_NEAREST), binwidth=1, 
                   fill="steelblue", color="white") +
    geom_rug(aes(x=seq(0, 15, 1)), color="firebrick")
plot(p2)

# Zoom in to find threshold
p3 <- p2 + xlim(c(0, 15)) + scale_y_sqrt()
plot(p3)
```

In this example, the unnormalized `hs5f` model distance threshold would be 
set to a value near 6.

## Calculate nearest neighbor distances across groups rather that within group

The arguments `fields` and `cross` are used for grouping data. 

`fields` will split the input `data.frame` in groups and will treat them
independently. It can be useful when the input data has multiple samples.

In the previous examples we used a subset of the original example data. In the
following example, we will use the two availabe samples, -1h and +7d, and will set `fields="SAMPLE"`. This will reproduce the same results as before for sample -1h.

```{r fields, eval=TRUE, warning=FALSE}
db <- ExampleDb
table(db$SAMPLE)
dist_hs1f_fields <- distToNearest(db, model="hs1f", first=FALSE, normalize="length", 
                           nproc=1, fields="SAMPLE")
library(testthat)
expect_equal(subset(dist_hs1f_fields, SAMPLE == "-1h"), dist_hs1f)
```

We can plot the nearest neighbor distances for the two samples:

```{r, eval=TRUE, warning=FALSE, fig.width=7}
# Generate m1n histogram
p4 <- ggplot() + theme_bw() + 
    ggtitle("Distance to nearest: hs1f") + xlab("distance") +
    geom_histogram(data=dist_hs1f_fields, aes(x=DIST_NEAREST), binwidth=0.025, 
                   fill="steelblue", color="white") +
    facet_grid( SAMPLE ~ ., scales = "free_y")
plot(p4)
```

Setting the argument `cross` allows to calculate the nearest neighbor distances 
across groups. In the following example we set `cross="SAMPLE"`. Data will be grouped
in two categories, being -1h and +7d. For each sequence, the distances to all sequences
belonging to other categories, +7d in this case, will be calculated and the 
smallest value returned.

```{r cross, eval=TRUE, warning=FALSE}
dist_hs1f_cross <- distToNearest(db, model="hs1f", first=FALSE, normalize="length", 
                           nproc=1, cross="SAMPLE")
```

The histogram for -1h will now be different, because we are looking at the distribution of
the distance between the sequences in -1h and +7d. 

```{r, eval=TRUE, warning=FALSE, fig.width=7}
# Generate m1n histogram
p5 <- ggplot() + theme_bw() + 
    ggtitle("Distance to nearest: hs1f, across samples") + xlab("distance") +
    geom_histogram(data=dist_hs1f_cross, aes(x=CROSS_DIST_NEAREST), binwidth=0.025, 
                   fill="steelblue", color="white") +
    facet_grid( SAMPLE ~ ., scales = "free_y")
plot(p5)
```


