---
title: "SHM: Tuning clonal assignment thresholds by calculating distance to nearest neighbor"
author: "Namita Gupta"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{Distance to nearest neighbor}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---
```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(dev='pdf')
```

Distance to nearest neighbor is calculated to estimate the threshold 
for partitioning clonally related sequences. This is done via the 
following steps:

1. Load a Change-O tab-delimited database file
2. Calculate distance to nearest neighbor
3. Generate histogram and inspect for threshold separating two modes

Load Change-O data
--------------------------------------------------------------------------------
A small example Change-O tab-delimited database file is included in the 
`alakazam` package. Distance to nearest neighbor requires the following fields 
(columns) to be present in the Change-O file: `V_CALL`, `J_CALL`, 
`JUNCTION_GAP_LENGTH`, and `JUNCTION`.

```{r, eval=TRUE, warning=FALSE, message=FALSE}
library(alakazam)
library(shm)

# Load Change-O file
file <- system.file("extdata", "Influenza.tab", package="shm")
db <- readChangeoDb(file)
```

Calculate distance to nearest neighbor
--------------------------------------------------------------------------------
The function for calculating distance between every sequence and its nearest
neighbor takes a few parameters to adjust how the distance is measured. If a 
genotype has been inferred using a tool (e.g., TIgGER) and the database has a
`V_CALL_GENOTYPED` field, then the parameter `genotyped` can be set to TRUE and 
the more accurate V call will be utilized. Furthermore, for a more leniency 
toward ambiguous segment calls, the parameter `first` can be set to FALSE, which 
would use the union of all possible genes to group sequences, rather than the 
first gene in the field. The `model` parameter determines which underlying SHM 
model is used to calculate the distance. The default model is `m3n`, a 
tri-nucleotide model from Smith et al., J Immunol, 1996. Other options include 
nucleotide (`ham`) and amino acid (`aa`) Hamming distance, a single nucleotide 
mouse model (`m1n`, Smith et al., J Immunol, 1996), and a 5-mer model inferred 
from human data (`hs5f`, Yaari et al., Front Immunol., 2013).


```{r, eval=TRUE, warning=FALSE}
# Calculate distance to nearest for genotyped sample data
db_hs5f <- distToNearest(df, vCallColumn="V_CALL_GENOTYPED", model="hs5f", first=FALSE)
```

Generate histogram
--------------------------------------------------------------------------------
The use of the distance to nearest calculation in the Change-O pipeline is to 
determine the optimal threshold for separating clonally related sequences 
(represented by sequences with 'near' neighbors) from singletons (sequences 
with no 'near' neighbor), which show up as two modes in a histogram. In the 
example provided here, this threshold is approximately at 0.002 for the 
S5F model.

```{r, eval=TRUE, warning=FALSE}
# Generate histogram
hist(db_dist$DIST_NEAREST, main='Histogram', xlab='Distance to Nearest', breaks=50)

# Zoom in to find threshold at around 2
hist(db_dist$DIST_NEAREST, main='Histogram - zoom', 
     xlab='Distance to Nearest', breaks=100, xlim=c(0,5))
```